### PR TITLE
[Kimi-K3] Add Kimi-K3 Vision Model

### PR DESCRIPTION
**Pull request by Xing-lil**
### PR Category

User Experience

### PR Types

Others

### Description

Add Kimi K3 Vision Model

### 是否引起精度变化

否

### CODE REVIEW

**Comment by liuhao2638** [comment]
@risemeup1111 review此pr

File: src/paddlefleet/models/kimi_k3/embedding.py [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+        self._check_inputs(pixel_values, grid_thws)
+
+        hidden_states = self.embedding(pixel_values, grid_thws)
+        rope_freqs_cis = self.rotary_pos_emb.get_freqs_cis(grid_thws)
+
+        return {
+            "hidden_states": hidden_states.unsqueeze(0),
+            "grid_thws": grid_thws,
+            "attention_mask": dict_args.get("attention_mask"),
+            "attn_mask_startend_row_indices": dict_args.get(
```
Comment (risemeup1111) [inline src/paddlefleet/models/kimi_k3/embedding.py]: P1

这里将 `attn_mask_startend_row_indices` 作为可选参数直接下传，却没有根据 `grid_thws` 构造各媒体的边界。调用方不传该参数时，BF16/FP16 会进入 `DotProductAttention` 的 `is_causal=True` 路径，使单张图像内部错误地变成因果注意力；float32 的 dense fallback 又不会隔离多张图像，导致跨图注意。当前测试均手工传入 mask，未覆盖真实的缺省路径。请在 vision tower 内基于 `grid_thws` 构造 packed/block-diagonal mask（float32 路径也需等价隔离），并补充缺省 mask 与 float32 多图隔离测试。

File: src/paddlefleet/models/kimi_k3/merge_input_ids.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    )
+    if batch_indices.shape[0] > 0:
+        image_to_overwrite = paddle.index_put(
+            image_to_overwrite,
+            (batch_indices, text_to_overwrite),
+            paddle.zeros([batch_indices.shape[0]], dtype="bool"),
+        )
+    pad_bound = (
+        paddle.cumsum(image_to_overwrite.astype("int64"), axis=-1) - 1
+    ) >= nb_image_pad.unsqueeze(1)
```
Comment (risemeup1111) [inline src/paddlefleet/models/kimi_k3/merge_input_ids.py]: P1

这里对所有样本统一用 `cumsum >= nb_image_pad` 去掉前部空位，只适用于左 padding。对于右 padding 的 `[8, IMG, 9, PAD]`（图片长度为 2），`nb_image_pad=1` 时未写入位置为 `[1, 2, 5]`，当前条件会保留 `[2, 5]`，使首个图片特征被放到 padding 之后，embedding、attention mask 和 position id 均错位。请按 padding side 分支：左 padding 保留当前下界，右 padding 排除尾部位置（例如 `< max_embed_dim - nb_image_pad`），并让测试断言精确的 embedding/mask 顺序。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
| 序号 | 位置 | 优先级 | 状态 |
| --- | --- | --- | --- |
| 1 | [embedding.py:273](https://github.com/PaddlePaddle/PaddleFleet/pull/1586#discussion_r3682665909) | P1 | ✅ |
| 2 | [merge_input_ids.py:153](https://github.com/PaddlePaddle/PaddleFleet/pull/1586#discussion_r3682667576) | P1 | ✅ |
| 3 | [PR 变更范围](https://github.com/PaddlePaddle/PaddleFleet/pull/1586#issuecomment-5140829424) | P3 | 🟡 |
| 4 | [PR 标题](https://github.com/PaddlePaddle/PaddleFleet/pull/1586#issuecomment-5140829424) | P3 | 🟡 |

本 PR 变更量较大（9 个文件 / 1996 行），建议按功能拆分：

- Vision tower：`embedding.py`、`kimi_k3_builders.py`、`kimi_k3_model.py`、`layer_specs.py`、`sd2_tpool_merge.py` 及视觉模型测试。
- Token merge utility：`merge_input_ids.py` 及对应测试。

标题建议（可直接复制）：`[New features] Add Kimi K3 Vision Model`

作者已明确选择保持当前变更范围和标题；以上两项作为非阻塞建议留档。

本轮按风险优先审查了视觉注意力掩码、图像/文本 token 合并、配置与构建链路及对应测试，未全量覆盖其余机械性代码。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Comment by Xing-lil** [comment]
@risemeup1111  已修改，重新review

File: src/paddlefleet/models/kimi_k3/embedding.py (hunk 1 above)
Comment (risemeup1111) [reply src/paddlefleet/models/kimi_k3/embedding.py]: 该提交已补上两个外部 mask 都为空时按 `grid_thws` 生成的 block-diagonal mask，但 `embedding.py:315` 仍要求两者同时为空才执行。若调用方只传 `attention_mask`，BF16/FP16 仍会缺少 start/end 边界；若只传 `attn_mask_startend_row_indices`，float32 dense 路径仍不会获得媒体隔离。当前测试也始终显式传入 start/end，未覆盖默认生成和 float32 路径。请在缺失任一掩码时补齐或合并内部媒体 mask，并增加这两种调用路径的回归测试。该 P1 仍保留。

File: src/paddlefleet/models/kimi_k3/merge_input_ids.py (hunk 2 above)
Comment (risemeup1111) [reply src/paddlefleet/models/kimi_k3/merge_input_ids.py]: 已核对本提交的左右 padding 分支及新增精确顺序断言：`slot_rank` 会按 padding side 排除尾部或前部空槽，右 padding 示例的 embedding、mask 与 position 顺序均已覆盖。该 P1 已修复。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

**Comment by Xing-lil** [comment]
@risemeup1111 已修改，重新review

File: src/paddlefleet/models/kimi_k3/embedding.py (hunk 1 above)
Comment (Xing-lil) [reply src/paddlefleet/models/kimi_k3/embedding.py]: 已修复

Comment (risemeup1111) [reply src/paddlefleet/models/kimi_k3/embedding.py]: 已核对最新提交：dense 分支现在会将基于 `grid_thws` 的 block-diagonal mask 与外部 mask 合并，BF16/FP16 分支在缺少 start/end 时也会补齐边界；新增回归测试覆盖了缺省、仅单一 mask 及 float32/BF16 的媒体隔离路径。该 P1 已修复。

**Comment by Xing-lil** [comment]
@risemeup1111 2个P3优先级的：3、PR 变更范围；4、PR 标题；均回复不同意，保持现状

**Review by risemeup1111 — APPROVED** [review]

**Review by From00 — APPROVED** [review]
LGTM
